wp-sw-cache/lib/service-worker.js

```diff
@@ -1,103 +1,127 @@
-(function(self){
-  var CACHE_PREFIX = '__wp-sw-cache::';
+(function(self, localforage){
 
-  var CACHE_NAME = CACHE_PREFIX + '$name';
-  var CACHE_FILES = $files;
+  // Represents the name of both storage and serviceworker
+  // This shouldn't change because the SW code below is smart enough to
+  // update what's needed and delete what isn't
+  var storageKey = '$name';
 
-  var debug = $debug;
+  // Outputs messages to the JS console if desired
+  var log = function() {
+    if($debug) {
+      console.log.apply(console, arguments);
+    }
+  };
 
-  self.addEventListener('install', function(event) {
-    // Perform install step:  loading each required file into cache
-    event.waitUntil(
-      caches.open(CACHE_NAME)
-        .then(function(cache) {
-          // Add all offline dependencies to the cache
-          if (debug) {
-            console.log('[install] Caches opened, adding all core components to cache');
-          }
-          return cache.addAll(CACHE_FILES);
-        })
-        .then(function() {
-          if (debug) {
-            console.log('[install] All required resources have been cached, we\'re good!');
-          }
-          return self.skipWaiting();
-        })
-    );
-  });
+  var wpSwCache = {
+    // Hash representing URLs and their modified times
+    // A changed hash will ensure the SW is downloaded when URLs change
+    version: '$version',
+    // A { url: hash } object which will be used to populate cache
+    urls: $urls,
+    // Instance of localForage to save urls and hashes to see if anything has changed
+    storage: localforage.createInstance({ name: storageKey }),
+    // Method to cleanse a URL before comparing and caching
+    normalizeAndAnonymize: function(request) {
+      var url = new URL(request.url);
+      if (url.origin !== location.origin) {
+        return request;
+      }
 
-  self.addEventListener('fetch', function(event) {
-    var request = event.request;
-    var lookupRequest = normalizeAndAnonymize(request);
-    if (!shouldBeHandled(lookupRequest)) {
-      return;
-    }
+      url.search = '';
+      url.fragment = '';
+      return new Request(url, {
+        method: request.method,
+        headers: request.headers,
+        mode: 'no-cors',
+        credentials: request.credentials,
+        cache: request.cache,
+        redirect: request.redirect,
+        referrer: request.referrer,
+        integrity: request.integrity
+      });
+    },
+    // Detect if a URL should be cacheable and in the desired URL list
+    shouldBeHandled: function(request) {
+      return request.method === 'GET' && (request.url in this.urls);
+    },
+    // Adds URLs to cache and localForage if the file has changed or needs to be added
+    update: function() {
+      // For every URL (file) the user wants cached...
+      return Promise.all(Object.keys(this.urls).map(function(url) {
```

Comment:
Wanted to be safe re: supporting service workers doesn't mean the browser supports arrow functions but you're probably right.